tests/unit/test_tls_lib.py

```diff
@@ -0,0 +1,311 @@
+
+        Note: this implicitly tests: _request_certificate & _parse_tls_file
+        """
+        # Tests for leader unit (ie internal certificates and external certificates)
+        self.harness.set_leader(True)
+        action_event = mock.Mock()
+        action_event.params = {}
+
+        # generated rsa key test - leader
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_internal_rsa_csr()
+        self.verify_external_rsa_csr()
+
+        # provided rsa key test - leader
+        set_app_rsa_key = self.harness.charm.app_peer_data["key"]
+        action_event.params = {"internal-key": set_app_rsa_key}
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_internal_rsa_csr(specific_rsa=True, expected_rsa=set_app_rsa_key)
+        self.verify_external_rsa_csr()
+
+        #  Tests for non-leader unit (ie external certificates)
+        self.harness.set_leader(False)
+        action_event = mock.Mock()
+        action_event.params = {}
+        app_rsa_key = self.harness.charm.app_peer_data["key"]
+        app_csr = self.harness.charm.app_peer_data["csr"]
+
+        # generated rsa key test - non-leader
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_external_rsa_csr()
+        # non-leaders should not reset the app key and app csr
+        self.verify_internal_rsa_csr(
+            specific_rsa=True, expected_rsa=app_rsa_key, specific_csr=True, expected_csr=app_csr
+        )
+
+        # provided rsa key test - non-leader
+        set_unit_rsa_key = self.harness.charm.unit_peer_data["key"]
+        action_event.params = {"external-key": set_unit_rsa_key}
+        self.harness.charm.tls._on_set_tls_private_key(action_event)
+        self.verify_external_rsa_csr(specific_rsa=True, expected_rsa=set_unit_rsa_key)
+        # non-leaders should not reset the app key and app csr
+        self.verify_internal_rsa_csr(
+            specific_rsa=True, expected_rsa=app_rsa_key, specific_csr=True, expected_csr=app_csr
+        )
+
+    @patch_network_get(private_address="1.1.1.1")
+    def test_tls_relation_joined_non_leader(self):
+        """Test that non-leader units set only external certificates."""
+        self.harness.set_leader(False)
+        self.relate_to_tls_certificates_operator()
+        # non leaders should not be allowed to set internal certificates
+        self.verify_internal_rsa_csr(
+            specific_rsa=True, expected_rsa=None, specific_csr=True, expected_csr=None
+        )
+        self.verify_external_rsa_csr()
+
+    @patch_network_get(private_address="1.1.1.1")
+    def test_tls_relation_joined_leader(self):
+        """Test that leader units set both external and internal certificates."""
+        self.harness.set_leader(True)
+        self.relate_to_tls_certificates_operator()
+        self.verify_internal_rsa_csr()
+        self.verify_external_rsa_csr()
```

Comment:
Opinion, I would split these tests into separate tests. 
Mostly for the reason that if a test fails, I'd rather know what was wrong from the test name and fix that, than have to debug it. 
It would add loads of lines of bloat for setting up the test scenario though, so (not necessary though) maybe you could add these tests into their own test class, each as methods, with a shared `setUp`?

Reply:
I totally agree, great suggestion, I found a way around the additional bloat :)